test/helpers/manifest.go

```diff
@@ -0,0 +1,272 @@
+	return &DeploymentManager{
+		deployments: map[string]*Deployment{},
+	}
+}
+
+// SetKubectl sets the kubectl client to use
+func (m *DeploymentManager) SetKubectl(kubectl *Kubectl) {
+	m.kubectl = kubectl
+}
+
+// DeployRandomNamespaceShared is like DeployRandomNamespace but will check if
+// the Manifest has already been deployed in any namespace. If so, returns the
+// namespace the existing deployment is running in. If not, the manifest is
+// deployed using DeployRandomNamespace.
+func (m *DeploymentManager) DeployRandomNamespaceShared(manifest Manifest) string {
+	if d, ok := m.deployments[manifest.Filename]; ok {
+		return string(d.namespace)
+	}
+
+	return m.DeployRandomNamespace(manifest)
+}
+
+// Deploy deploys a manifest into a random namespace using the deployment
+// manager and stores the deployment in the manager
+func (m *DeploymentManager) DeployRandomNamespace(manifest Manifest) string {
+	namespace := GenerateNamespaceForTest("")
+
+	res := m.kubectl.NamespaceCreate(namespace)
+	if !res.WasSuccessful() {
+		ginkgoext.Failf("Unable to create namespace %s: %s",
+			namespace, res.OutputPrettyPrint())
+	}
+
+	d, err := manifest.deploy(m.kubectl, namespace)
+	if err != nil {
+		ginkgoext.By("Deleting namespace %s", namespace)
+		m.kubectl.NamespaceDelete(namespace)
+
+		ginkgoext.Failf("Unable to deploy manifest %s: %s", manifest.Filename, err)
+	}
+
+	m.deployments[manifest.Filename] = d
+
+	return namespace
+}
+
+// Deploy deploys a manifest using the deployment manager and stores the
+// deployment in the manager
+func (m *DeploymentManager) Deploy(namespace string, manifest Manifest) {
+	d, err := manifest.deploy(m.kubectl, namespace)
+	if err != nil {
+		ginkgoext.Failf("Unable to deploy manifest %s: %s", manifest.Filename, err)
+	}
+
+	m.deployments[manifest.Filename] = d
+}
+
+// DeleteAll deletes all deployments which have previously been deployed using
+// this deployment manager
+func (m *DeploymentManager) DeleteAll() {
+	var (
+		deleted = 0
+		wg      sync.WaitGroup
+	)
+
+	wg.Add(len(m.deployments))
+	for _, d := range m.deployments {
+		// Issue all delete triggers in parallel
+		go func(d *Deployment) {
+			d.Delete()
+			wg.Done()
+		}(d)
+		deleted++
+	}
+
+	namespaces := map[NamespaceName]struct{}{}
+	for _, d := range m.deployments {
+		if d.namespace.IsRandom() {
+			namespaces[d.namespace] = struct{}{}
+		}
+	}
+
+	wg.Wait()
+	m.deployments = map[string]*Deployment{}
+
+	wg.Add(len(namespaces))
+	for namespace := range namespaces {
+		ginkgoext.By("Deleting namespace %s", namespace)
+		go func(namespace NamespaceName) {
+			m.kubectl.NamespaceDelete(string(namespace))
+			wg.Done()
+		}(namespace)
+	}
+
+	if deleted > 0 {
+		m.kubectl.WaitCleanAllTerminatingPods(2 * time.Minute)
+	}
+	wg.Wait()
+}
+
+// DeleteCilium deletes a Cilium deployment that was previously deployed with
+// DeployCilium()
+func (m *DeploymentManager) DeleteCilium() {
+	if m.ciliumDeployed {
+		m.kubectl.DeleteCiliumDS()
```

Comment:
While most tests seem to use `kubectl.DeleteCiliumDS()`, I've found that it is not always sufficient. For example, if one deploys Cilium with Helm options that spawn additional pods (e.g. `global.hubble.cli.enabled=true` which spawns a bunch a DaemonSet for running the Hubble CLI), then `DeleteCiliumDS()` will not delete these additionally create resources.

While I'm not very familiar with what most tests expect and don't know why tests are using `DeleteCiliumDS` in the first place. But I believe the correct thing to do here is to to actually invoke something like `m.kubectl.Delete(ciliumDeployment.path)`, i.e. delete all resources created in `DeployCilium`.


Other than that, I really like the idea overall idea of this PR. It allows us to be more consistent.

Reply:
Good point. I've used what k8sT/DatapathConfiguration.go has used but I agree that deleting the original deployment makes more sense.